sympy/ntheory/partitions_.py

```diff
@@ -1,40 +1,101 @@
 from __future__ import print_function, division
-
 from mpmath.libmp import (fzero,
     from_man_exp, from_int, from_rational,
     fone, fhalf, bitcount, to_int, to_str, mpf_mul, mpf_div, mpf_sub,
-    mpf_add, mpf_sqrt, mpf_pi, mpf_cosh_sinh, pi_fixed, mpf_cos)
+    mpf_add, mpf_sqrt, mpf_pi, mpf_cosh_sinh, pi_fixed, mpf_cos, mpf_sin)
 from sympy.core.numbers import igcd
 import math
 from sympy.core.compatibility import range
+from .residue_ntheory import (_sqrt_mod_prime_power,
+    legendre_symbol, jacobi_symbol, is_quad_residue)
+
+def _pre():
+    maxn = 10**5
+    global _factor
+    global _totient
+    _factor = [0]*maxn
+    _totient = [1]*maxn
+    lim = int(maxn**0.5) + 5
+    for i in range(2, lim):
+        if _factor[i] == 0:
+            for j in range(i*i, maxn, i):
+                if _factor[j] == 0:
+                    _factor[j] = i
+    for i in range(2, maxn):
+        if _factor[i] == 0:
+            _factor[i] = i
+            _totient[i] = i-1
+            continue
+        x = _factor[i]
+        y = i//x
+        if y%x == 0:
+            _totient[i] = _totient[y]*x
+        else:
+            _totient[i] = _totient[y]*(x - 1)
 
+def _a(n, k, prec):
+    """ Compute the inner sum in HRR formula
+    References
+    ==========
 
-def _a(n, j, prec):
-    """Compute the inner sum in the HRR formula."""
-    if j == 1:
+    - http://msp.org/pjm/1956/6-1/pjm-v6-n1-p18-p.pdf
+
+    """
+    if k == 1:
         return fone
-    s = fzero
-    pi = pi_fixed(prec)
-    for h in range(1, j):
-        if igcd(h, j) != 1:
-            continue
-        # & with mask to compute fractional part of fixed-point number
-        one = 1 << prec
-        onemask = one - 1
-        half = one >> 1
-        g = 0
-        if j >= 3:
-            for k in range(1, j):
-                t = h*k*one//j
-                if t > 0:
-                    frac = t & onemask
-                else:
-                    frac = -((-t) & onemask)
-                g += k*(frac - half)
-        g = ((g - 2*h*n*one)*pi//j) >> prec
-        s = mpf_add(s, mpf_cos(from_man_exp(g, -prec), prec), prec)
-    return s
 
+    k1 = k
+    e = 0
+    p = _factor[k]
+    while k1%p == 0:
+        k1 //= p
+        e += 1
+    k2 = k//k1 # k2 = p^e
+    v = 1 - 24*n
+    pi = mpf_pi(prec)
+
+    if k1 == 1:
+        # k  = p^e
+        if p == 2:
+            mod = 8*k
+            v = (mod + v%mod)%mod
```

Comment:
I thought `v < mod` would be safer for passing to `_sqrt_mod_prime_power`, because I hadn't read its implementation. Also, it doesn't affect the time taken at all. So, I don't think there is a need to change it.